yksgi/src/llvmbridge.rs

```diff
@@ -0,0 +1,210 @@
+    }
+}
+
+pub struct LLVMBasicBlock(pub LLVMBasicBlockRef);
+
+impl LLVMBasicBlock {
+    pub unsafe fn new(bb: LLVMBasicBlockRef) -> Self {
+        Self(bb)
+    }
+
+    pub unsafe fn first(&self) -> LLVMValue {
+        self.instruction(0)
+    }
+
+    pub unsafe fn instruction(&self, instridx: usize) -> LLVMValue {
+        let mut instr = LLVMGetFirstInstruction(self.0);
+        for _ in 0..instridx {
+            instr = LLVMGetNextInstruction(instr);
+        }
+        LLVMValue::new(instr)
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Hash)]
+pub struct LLVMType(pub LLVMTypeRef);
+impl LLVMType {
+
+    pub unsafe fn kind(&self) -> LLVMTypeKind {
+        LLVMGetTypeKind(self.0)
+    }
+
+    pub unsafe fn is_pointer(&self) -> bool {
+        matches!(self.kind(), LLVMTypeKind::LLVMPointerTypeKind)
+    }
+
+    pub unsafe fn is_struct(&self) -> bool {
+        matches!(self.kind(), LLVMTypeKind::LLVMStructTypeKind)
+    }
+
+    pub unsafe fn as_str(&self) -> &CStr {
+        CStr::from_ptr(LLVMPrintTypeToString(self.0))
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Copy)]
+pub struct LLVMValue(pub LLVMValueRef);
+impl LLVMValue {
+
+    pub unsafe fn new(vref: LLVMValueRef) -> Self {
+        LLVMValue(vref)
+    }
+
+    pub unsafe fn is_constant(&self) -> bool {
+        !LLVMIsAConstant(self.0).is_null()
+    }
+
+    pub unsafe fn is_br(&self) -> bool {
+        !LLVMIsABranchInst(self.0).is_null()
+    }
+
+    pub unsafe fn is_gep(&self) -> bool {
+        !LLVMIsAGetElementPtrInst(self.0).is_null()
+    }
+
+    pub unsafe fn is_instruction(&self) -> bool {
+        !LLVMIsAInstruction(self.0).is_null()
+    }
+
+    pub unsafe fn is_load(&self) -> bool {
+        !LLVMIsALoadInst(self.0).is_null()
+    }
+
+    pub unsafe fn is_store(&self) -> bool {
+        !LLVMIsAStoreInst(self.0).is_null()
+    }
+
+    pub unsafe fn is_switch(&self) -> bool {
+        !LLVMIsASwitchInst(self.0).is_null()
+    }
+
+    pub unsafe fn get_type(&self) -> LLVMType {
+        LLVMType(LLVMTypeOf(self.0))
+    }
+
+    pub unsafe fn get_operand(&self, idx: u32) -> LLVMValue {
+        debug_assert!(!LLVMIsAUser(self.0).is_null());
+        LLVMValue(LLVMGetOperand(self.0, idx))
+    }
+
+    pub unsafe fn opcode(&self) -> LLVMOpcode {
+        debug_assert!(!LLVMIsAInstruction(self.0).is_null());
+        LLVMGetInstructionOpcode(self.0)
+    }
+
+    pub unsafe fn as_str(&self) -> &CStr {
+        CStr::from_ptr(LLVMPrintValueToString(self.0))
+    }
+}
+
+pub unsafe fn llvm_const_to_sgvalue(c: LLVMValueRef) -> SGValue {
+    let ty = LLVMTypeOf(c);
+    let kind = LLVMGetTypeKind(ty);
+    match kind {
+        LLVMTypeKind::LLVMIntegerTypeKind => {
+            let _width = LLVMGetIntTypeWidth(ty);
+            let val = LLVMConstIntGetZExtValue(c) as u64;
+            SGValue::new(val, LLVMType(ty))
+        }
+        _ => todo!(),
+    }
+}
+
+/// Get the original instructions of a live variable before it is being
```

Comment:
nitpick: I'd re-word this comment, as it reads a little clumsily.

Reply:
Fixed in ffd4e75